wellcomeml/ml/keras_utils.py

```diff
@@ -16,9 +38,26 @@ def on_epoch_end(self, epoch, logs):
         f1 = round(f1_score(Y_val, Y_pred, average="micro"), 4)
         p = round(precision_score(Y_val, Y_pred, average="micro"), 4)
         r = round(recall_score(Y_val, Y_pred, average="micro"), 4)
+
+        self.f1s.append(f1)
+        self.recalls.append(r)
+        self.precisions.append(p)
+
         print(f" - val metrics: P {p:.4f} R {r:.4f} F {f1:.4f}")
+
         return
 
+    def on_train_end(self, logs={}):
+        """Write metrics to csv file"""
+
+        if self.history_path:
+            history = {}
+            history["f1"] = self.f1s
+            history["precision"] = self.precisions
+            history["recall"] = self.recalls
+            history_df = pd.DataFrame(history)
+            history_df.to_csv(self.history_path, index_label="epoch")
```

Comment:
Pandas seems a bit of a cannon to kill a fly here. Can we save in jsonL?

Reply:
Yeah this is true, and I had been thinking the same. Ideally we do want it in csv so that it can be easily read by dvc (the jsonl structure  required for this is a bit convoluted), let me switch this out and use the csv module instead.